### PR TITLE
Tool box react version

### PR DESCRIPTION
**Pull request by Ivylyu**
add a toolbox under tabs

### CODE REVIEW

**Comment by whimet** [comment]
冲突了，不要提交bundle.js

**Comment by MrCoder** [comment]
> 冲突了，不要提交bundle.js
之前都是提交bundle.js的。如果修改这个规则我觉得不要在吕晔的PR上改吧。

**Comment by Ivylyu** [comment]
> 冲突了，不要提交bundle.js

那我需要把这个文件放到.git里然后再提交一下吗？我看它告诉我conflict的文件是app.jsx这个需要改吗？

File: src/components/ContentWrap.jsx [hunk 1]
```diff
@@ -687,6 +689,34 @@ export default class ContentWrap extends Component {
 		}
 	}
 
+	toolboxUpdateToApp(param) {
+		if (param === "NewParticipant"){
+			this.addNewParticipant();
+		} else {
+			this.cmCodes.js === '' ? this.cm.js.setValue(param): this.cm.js.setValue(`${this.cmCodes.js}\n${param}`);
```
Comment (whimet) [inline src/components/ContentWrap.jsx]: we should separate the code calculation logic and rendering logic; same applies on `addNewParticipant`

File: src/components/app.jsx [hunk 2]
```diff
@@ -1487,7 +1488,7 @@ export default class App extends Component {
 					<input
 						type="hidden"
 						name="data"
-						value='{"title": "New Pen!", "html": "<div>Hello, World!</div>"}'
```
Comment (whimet) [inline src/components/app.jsx]: why do we need this change?

File: src/components/app.jsx [hunk 3]
```diff
@@ -1360,7 +1361,7 @@ export default class App extends Component {
...
+						value="{&quot;title&quot;: &quot;New Pen!&quot;, &quot;html&quot;: &quot;<div>Hello, World!</div>&quot;}"
```
Comment (whimet) [inline src/components/app.jsx]: why do we need this change?

File: src/components/ContentWrap.jsx (hunk 1 above)
Comment (Ivylyu) [reply src/components/ContentWrap.jsx]: what do you mean by logic and rendering logic?

Comment (Ivylyu) [reply src/components/ContentWrap.jsx]: Other changes were not by me

File: package.json [hunk 4]
```diff
@@ -71,6 +71,7 @@
 	},
 	"dependencies": {
 		"@emmetio/codemirror-plugin": "^0.5.4",
+		"@firebase/firestore": "^1.0.5",
```
[the remark this answers is not in the document]
Comment (MrCoder) [reply package.json]: Removed. I was getting errors for firebase versions. It looks fine after I removed this line.

File: src/components/app.jsx (hunk 3 above)
Comment (MrCoder) [reply src/components/app.jsx]: Reverted.

File: src/components/ContentWrap.jsx (hunk 1 above)
Comment (MrCoder) [reply src/components/ContentWrap.jsx]: Note that the method `this.cm.js.setValue` is called three times - twice in `toolboxUpdateToApp` and once in `addNewParticipant`. This is a [code smell](https://en.wikipedia.org/wiki/Code_smell).

Yanhui means we could have one method to generate the code only and set js value in ONE place. This way we have less duplicate code. This is often called DRY (don't repeat yourself). I will make this change.

Comment (MrCoder) [reply src/components/ContentWrap.jsx]: @whimet 我改了，在计算code的地方还是有些重复。明天再想想有没有好的算法。

Comment (Ivylyu) [reply src/components/ContentWrap.jsx]: ok, now I know what we are talking about

File: src/components/app.jsx (hunk 2 above)
Comment (MrCoder) [reply src/components/app.jsx]: reverted.

**Pull request merged**